platform/cms/accelerated_publishing/rfc/reverse-proxy-falback.md

```diff
@@ -0,0 +1,93 @@
+# RFC: Content Proxy Location Fallback
+
+- Date: 2024-01-03
+- Related Issue: [va.gov-cms #14469](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/14469) (Content Build / Next Build traffic routing)
+
+## Background
+The changes proposed in this RFC are designed to address bringing [Next Build / Accelerated Publishing](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/6577) to production. Next Build is a new front-end build system intended to replace [Content Build](https://github.com/department-of-veterans-affairs/content-build) as the mechanism by which CMS content is built delivered to the Veteran.
+
+Currently VA.gov uses a single S3 bucket as storage for the output of Content Build and the source for content files that are served to web requests. [Traffic is relayed by the reverse proxy to this S3 bucket](https://github.com/department-of-veterans-affairs/vsp-platform-revproxy/blob/main/template-rendering/revproxy-vagov/templates/nginx_website_server.conf.j2#L393).
+
+This [VA EKS architecture document](https://vfs.atlassian.net/wiki/spaces/OT/pages/2231304195/Platform+Infrastructure+Diagrams#EKS-Architecture) is a good high-level overview of how traffic flow works. We are specifically addressing traffic to 'Website' in this proposal:
+
+![Current state of traffic flow](assets/revproxy_currentstate.png)
+
+We are not proposing any changes to API traffic handling. Only web content requests would be affected.
+
+## Motivation
+Content publishing will migrate from Content Build to Next Build gradually, as products are implemented in Next Build and able to be published there. It is unlikely that we could migrate every type of content currently built by Content Build prior to taking Next Build to production. Coexistance of the two systems will be necessary during the migration and rollout.
+
+This rollout will be greatly simplified if each system publishes to its own S3 bucket, rather than to the same bucket (reasons for this are given in the Alternatives section below). However, web requests will need to be routed to the appropriate S3 bucket.
+
+We propose an addition to the reverse proxy which would direct web traffic to the new Next Build bucket first. If content is available at that bucket, it is served. If a 4xx/5xx response is returned, the reverse proxy would then direct traffic to the Content Build server.
+
+## Design
+This diagram shows roughly what we propose: ![Diagram showing updated infrastructure](assets/revproxfallback.drawio.png)
+
+At an nginx level, this is accomplished by adding `proxy_intercept_errors on;` to the primary location configuration for web content and adding a named fallback location as the handler for those errors. The primary `proxy_pass` location would point to the Next Build S3 bucket, and the fallback location `proxy_pass` would point to the Content Build S3 bucket.
```

Comment:
Just to be annoyingly technical, `proxy_intercept_errors on;` is already in the revproxy config, part of the proxy_pass configuration used for sending requests to the content-build bucket. We'd effectively be changing how the config handles the 404s it intercepts and instead of invoking a 404.html page, go our custom fallback instead. Just a point of clarity.

Reply:
That's not annoyingly technical, it's precise and an important part of what we're doing. Thank you for catching it.